File: shared/notify-prefs.ts
import type { NotificationKind } from "./notify.ts";

/**
 * What this phone wants to be told about.
 *
 * The column has existed since push landed and nothing has ever read it: `prefs` was stored on the
 * token, parsed back out, and then every message went to every device anyway. So a switch built
 * against it would have looked like it worked and done nothing — which is worse than not having
 * one. This is the shape, and `allows` is the single place that decides.
 *
 * Two levels, because a household needs both. A device says which *kinds* it cares about — plenty
 * of people want the Sunday results and not the Thursday nudge — and then, separately, which
 * *entries* it is hearing about at all. A parent running four entries does not want four phones
 * buzzing four times; they want their own, and the reminder for whoever has not picked.
 *
 * **Absent always means on.** A new kind ships enabled for everybody who never touched a switch,
 * an unknown key from an older build is ignored rather than muting something, and an empty object
 * is the default rather than silence. Getting that backwards would mean a Worker deploy silently
 * switching off a notification somebody was relying on.
 */

export interface EntryPrefs {
  /** Mutes this entry on this device entirely, whatever the kinds say. */
  muted?: boolean;
  /** Per-kind overrides for this one entry. Absent falls through to the device default. */
  kinds?: Partial<Record<NotificationKind, boolean>>;
}

export interface NotifyPrefs {
  /** The device's default per kind. Absent means on. */
  kinds?: Partial<Record<NotificationKind, boolean>>;
  /** Overrides for particular entries, keyed by player id. */
  entries?: Record<string, EntryPrefs>;
}

export const NOTIFICATION_KINDS: NotificationKind[] = ["picksDue", "segment", "weekDone"];

/** What each switch says on the screen, so the wording lives with the model rather than in a view. */
export const KIND_LABELS: Record<NotificationKind, { title: string; detail: string }> = {
  picksDue: {
    title: "Pick reminders",
    detail: "Before the first game, and again before Sunday, if your five are not in.",
  },
  segment: {
    title: "When your games finish",
    detail: "One message per slate you had a pick in — not one per game.",
  },
  weekDone: {
    title: "The week's result",
    detail: "Where you finished, and where that leaves your season.",
  },
};

function isKind(value: string): value is NotificationKind {
  return (NOTIFICATION_KINDS as string[]).includes(value);
}

/**
 * Reads whatever is in the column, including what older builds wrote.
 *
 * Push shipped storing a flat `{ picksDue: false }` and this takes that to mean the device
 * default, which is what it meant. Anything unrecognised is dropped rather than guessed at: a
 * malformed blob has to mean "everything on", because the alternative is a phone that has gone
 * quiet and no way for its owner to find out why.
 */
export function parsePrefs(raw: unknown): NotifyPrefs {
  if (!raw || typeof raw !== "object" || Array.isArray(raw)) return {};
  const source = raw as Record<string, unknown>;
  const prefs: NotifyPrefs = {};

  const kinds = readKinds(source.kinds);
  // The flat shape the first version wrote, read as the device default it always was.
  const flat = readKinds(source);
  const merged = { ...flat, ...kinds };
  if (Object.keys(merged).length) prefs.kinds = merged;

  if (source.entries && typeof source.entries === "object" && !Array.isArray(source.entries)) {
    const entries: Record<string, EntryPrefs> = {};
    for (const [id, value] of Object.entries(source.entries as Record<string, unknown>)) {
      if (!value || typeof value !== "object" || Array.isArray(value)) continue;
      const entry = value as Record<string, unknown>;
      const own: EntryPrefs = {};
      if (entry.muted === true) own.muted = true;
      const entryKinds = readKinds(entry.kinds);
      if (Object.keys(entryKinds).length) own.kinds = entryKinds;
      if (own.muted || own.kinds) entries[id] = own;
    }
    if (Object.keys(entries).length) prefs.entries = entries;
  }

  return prefs;
}

/** Only booleans, only known kinds, and only `false` is worth storing — absent means on. */
function readKinds(raw: unknown): Partial<Record<NotificationKind, boolean>> {
  if (!raw || typeof raw !== "object" || Array.isArray(raw)) return {};
  const out: Partial<Record<NotificationKind, boolean>> = {};
  for (const [key, value] of Object.entries(raw as Record<string, unknown>)) {
    if (typeof value === "boolean" && isKind(key)) out[key] = value;
  }
  return out;
}

/**
 * Should this device hear this message about this entry?
 *
 * The order is most specific first: an entry that is muted hears nothing, then that entry's own
 * answer for this kind, then the device's default for the kind, then yes.
 */
export function allows(prefs: NotifyPrefs, kind: NotificationKind, entryId: string): boolean {
  const entry = prefs.entries?.[entryId];
  if (entry?.muted) return false;
  const own = entry?.kinds?.[kind];
  if (own !== undefined) return own;
  const device = prefs.kinds?.[kind];
  if (device !== undefined) return device;
  return true;
}

/** Everything that is switched off, for a settings screen to draw without reimplementing the rule. */
export function isMuted(prefs: NotifyPrefs, entryId: string): boolean {
  return prefs.entries?.[entryId]?.muted === true;
}

File: ios/TallyKit/Sources/TallyKit/Models/NotifyPrefs.swift
import Foundation

/**
 What this phone wants to be told about, mirroring `shared/notify-prefs.ts`.

 Two levels, because a household needs both. A device says which *kinds* it cares about — plenty of
 people want the Sunday results and not the Thursday nudge — and then, separately, which *entries*
 it hears about at all. A parent running four entries does not want four phones buzzing four times.

 **Absent always means on.** A new kind ships enabled for everybody who never touched a switch, an
 unknown key from an older build is ignored rather than muting something, and an empty object is
 the default rather than silence. The opposite would mean a deploy quietly switching off a
 notification somebody was relying on, with nothing on screen to say why.
 */
public enum NotificationKind: String, Codable, CaseIterable, Sendable {
    case picksDue
    case segment
    case weekDone

    public var title: String {
        switch self {
        case .picksDue: "Pick reminders"
        case .segment: "When your games finish"
        case .weekDone: "The week's result"
        }
    }

    public var detail: String {
        switch self {
        case .picksDue: "Before the first game, and again before Sunday, if your five are not in."
        case .segment: "One message per slate you had a pick in — not one per game."
        case .weekDone: "Where you finished, and where that leaves your season."
        }
    }
}

public struct EntryNotifyPrefs: Codable, Hashable, Sendable {
    /// Mutes this entry on this device entirely, whatever the kinds say.
    public var muted: Bool?
    /// Per-kind overrides for this one entry. Absent falls through to the device default.
    public var kinds: [String: Bool]?

    public init(muted: Bool? = nil, kinds: [String: Bool]? = nil) {
        self.muted = muted
        self.kinds = kinds
    }
}

public struct NotifyPrefs: Codable, Hashable, Sendable {
    /// The device's default per kind. Absent means on.
    public var kinds: [String: Bool]?
    /// Overrides for particular entries, keyed by player id.
    public var entries: [String: EntryNotifyPrefs]?

    public init(kinds: [String: Bool]? = nil, entries: [String: EntryNotifyPrefs]? = nil) {
        self.kinds = kinds
        self.entries = entries
    }

    public static let everything = NotifyPrefs()

    // MARK: Reading

    /// The device default for a kind. The same rule the server applies, so a switch on screen and
    /// the message that does or does not arrive can never disagree.
    public func isOn(_ kind: NotificationKind) -> Bool {
        kinds?[kind.rawValue] ?? true
    }

    public func isMuted(entry: String) -> Bool {
        entries?[entry]?.muted == true
    }

    /// Whether this entry hears this kind, which is the whole rule in one place: a muted entry
    /// hears nothing, then that entry's own answer, then the device's, then yes.
    public func allows(_ kind: NotificationKind, entry: String) -> Bool {
        if isMuted(entry: entry) { return false }
        if let own = entries?[entry]?.kinds?[kind.rawValue] { return own }
        return isOn(kind)
    }

    // MARK: Writing

    /// Only `false` is stored. Turning something back on removes the key rather than writing
    /// `true`, so the stored object stays the set of deliberate exceptions and a kind added later
    /// arrives switched on.
    public mutating func set(_ kind: NotificationKind, on: Bool) {
        var next = kinds ?? [:]
        if on { next.removeValue(forKey: kind.rawValue) } else { next[kind.rawValue] = false }
        kinds = next.isEmpty ? nil : next
    }

    public mutating func setMuted(_ muted: Bool, entry: String) {
        var all = entries ?? [:]
        var own = all[entry] ?? EntryNotifyPrefs()
        own.muted = muted ? true : nil
        all[entry] = own.isEmpty ? nil : own
        entries = all.isEmpty ? nil : all
    }

    public mutating func set(_ kind: NotificationKind, on: Bool, entry: String) {
        var all = entries ?? [:]
        var own = all[entry] ?? EntryNotifyPrefs()
        var ownKinds = own.kinds ?? [:]
        // An entry-level switch that agrees with the device default is not an override, so it is
        // removed: the entry follows the device again, including when the device changes later.
        if on == isOn(kind) {
            ownKinds.removeValue(forKey: kind.rawValue)
        } else {
            ownKinds[kind.rawValue] = on
        }
        own.kinds = ownKinds.isEmpty ? nil : ownKinds
        all[entry] = own.isEmpty ? nil : own
        entries = all.isEmpty ? nil : all
    }
}

extension EntryNotifyPrefs {
    /// Nothing worth storing — no mute and no overrides.
    var isEmpty: Bool { muted != true && (kinds?.isEmpty ?? true) }
}
